gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java

```diff
@@ -174,39 +190,156 @@ private Path resolvePath(String pattern) {
 
+          Partition partition = partitionEntry.getValue();
           InputFormat<?, ?> inputFormat = getInputFormat(partition.getTPartition().getSd());
 
-          for (CopyableFile.Builder builder : getCopyableFilesFromPaths(getPaths(inputFormat, partition.getLocation()),
+          Optional<Partition> targetPartition = Optional.fromNullable(targetPartitions.get(partitionEntry.getKey()));
+
+          if (targetPartition.isPresent()) {
+            targetPartitions.remove(partitionEntry.getKey());
+            try {
+              checkPartitionCompatibility(partition, targetPartition.get());
+            } catch (IOException ioe) {
+              log.warn("Source and target partitions are not compatible. Will override target partition.");
+              // TODO: add partition deregistration step
+              targetPartition = Optional.absent();
+            }
+          }
+
+          Collection<Path> sourcePaths = getPaths(inputFormat, partition.getLocation());
+          Collection<Path> targetExistingPaths = targetPartition.isPresent() ?
+              getPaths(inputFormat, targetPartition.get().getLocation()) : Sets.<Path>newHashSet();
+
+          DiffPathSet diffPathSet =
+              diffSourceAndTargetPaths(sourcePaths, targetExistingPaths, Optional.of(partition), targetFs);
+
+          // TODO: create delete work units for files that must be deleted on target
+
+          for (CopyableFile.Builder builder : getCopyableFilesFromPaths(diffPathSet.filesToCopy,
               configuration, Optional.of(partition))) {
             copyableFiles.add(builder.fileSet(gson.toJson(partition.getValues())).build());
           }
 
+          if (!targetPartition.isPresent()) {
+            HiveSpec partitionHiveSpec = new SimpleHiveSpec.Builder(this.table.getDataLocation()).
+                withTable(this.table.getTTable()).withPartition(Optional.of(partition.getTPartition())).build();
+            // TODO: add partition registration step
+          }
+
         }
+
+        // Partitions that don't exist in source
+        for (Map.Entry<List<String>, Partition> partitionEntry : targetPartitions.entrySet()) {
+          // TODO: add partition deregistration step
+        }
+
       } else {
         InputFormat<?, ?> inputFormat = getInputFormat(this.table.getSd());
-        for (CopyableFile.Builder builder : getCopyableFilesFromPaths(getPaths(inputFormat, this.table.getSd().getLocation()),
+
+        Collection<Path> sourcePaths = getPaths(inputFormat, this.table.getSd().getLocation());
+        Collection<Path> targetExistingPaths = targetTable.isPresent() ?
+            getPaths(inputFormat, targetTable.get().getSd().getLocation()) : Sets.<Path>newHashSet();
+
+        DiffPathSet diffPathSet = diffSourceAndTargetPaths(sourcePaths, targetExistingPaths,
+            Optional.<Partition>absent(), targetFs);
+
+        // TODO: create delete work units for files that must be deleted on target
```

Comment:
How will the delete-workunits fit here? The return type is `copyableFile`, not `copyableEntity`.